File: python/LargeSelfFamilyRef.py
from __future__ import annotations

# coding: utf-8
# LargeSelfFamilyRef.py

from itertools import *
from collections import defaultdict
import sys

from typing import Optional

from VCF import *
from pedigree import PedigreeTable, Family
from KnownFamily import KnownFamily
from GenoRecord import GenoRecord
from VCFGeno import VCFGenoBase, VCFGeno
from VCFSelfImputable import *
from VCFSelfParentImputed import VCFSelfParentImputed
from VCFSelfProgenyImputed import VCFSelfProgenyImputed
from VCFSelfImputable import *
import RefCommon
from Map import *
from option import *
from Genotype import Genotype
from common import *


#################### LargeSelfFamilyRef ####################

def create_parent_phased_vcf(family: KnownFamily, orig_vcf: VCFSmall,
								phased_vcf: VCFGeno, gmap: Map,
								op: Option) -> Optional[VCFSelfParentImputed]:
	# phasedの後代を取り除いたsamplesを作る
	ids = phased_vcf.extract_columns(family.samples())	# phased_vcf内のindex
	prog_ids = [ i for i in range(1, len(family.samples())) if ids[i] == -1 ]
	if len(prog_ids) == 0:
		return None
	
	samples = [family.mat] + [ family.samples()[i] for i in prog_ids ]
	vcf_progs = VCFGeno.extract_samples(samples[1:], orig_vcf)
	records = RefCommon.merge_records(phased_vcf, vcf_progs, samples)
	return VCFSelfParentImputed(samples, records, gmap, phased_vcf.vcf)

def create_progeny_phased_vcf(family: KnownFamily, orig_vcf: VCFSmall,
								ref_vcf: VCFGeno, phased_vcf: VCFGeno,
								gmap: Map, op: Option) -> VCFSelfProgenyImputed:
	# phasedの後代を一つだけ残したsamplesを作る
	N = len(family.samples())
	ids = phased_vcf.extract_columns(family.samples())	# phased_vcf内のindex
	prog_ids = [ i for i in range(1, len(family.samples())) if ids[i] != -1 ]
	phased_prog_id = prog_ids[0]
	phased_sample = family.samples()[phased_prog_id]
	set_prog_ids = set(prog_ids[1:])	# 最初の1個だけ残す
	samples = [ family.samples()[i] for i in range(N) if i not in set_prog_ids ]
	non_phased_samples = [ s for s in samples if s != phased_sample ]
	
	non_phased_vcf = VCFGeno.extract_samples(non_phased_samples, orig_vcf)
	records = RefCommon.merge_records(phased_vcf, non_phased_vcf, samples)
	ref_haps = ref_vcf.create_ref_haps()
	return VCFSelfProgenyImputed(samples, records, ref_haps,
										prog_ids[0], gmap, phased_vcf.vcf)

def impute(families: list[KnownFamily], orig_vcf: VCFSmall,
								merged_vcf: Optional[VCFGeno],
								ref_vcf: VCFGeno,
								gmap: Map, op: Option) -> Optional[VCFGeno]:
	if not families:
		return None
	
	# 今までimputeしたVCFとリファレンスを合わせてphasedという
	phased_vcf = VCFGeno.merge(merged_vcf, ref_vcf)
	
	vcfs: list[VCFGenoBase] = []
	set_phased = set(phased_vcf.samples)
	for family in families:
		if family.mat in set_phased:
			# 親がphasedならば、そのままimputeする
			vcf1 = create_parent_phased_vcf(family, orig_vcf,
												phased_vcf, gmap, op)
			if vcf1 is not None:
				vcf1.impute()
				vcfs.append(vcf1)
		elif set_phased & set(family.samples()):
			# 後代にphasedがあれば、そのままimputeする
			vcf2 = create_progeny_phased_vcf(family, orig_vcf, ref_vcf,
														phased_vcf, gmap, op)
			vcf2.impute()
			vcfs.append(vcf2)
	
	if merged_vcf is not None:
		vcfs.append(merged_vcf)
	
	if vcfs:
		return VCFGeno.join(vcfs, orig_vcf.samples)
	else:
		return None

__all__ = ['impute']

File: src/LargeSelfFamilyRef.cpp
#include "../include/VCFGeno.h"
#include "../include/LargeSelfFamilyRef.h"
#include "../include/VCFSelfFillable.h"
#include "../include/VCFSelfImputable.h"
#include "../include/VCFSelfParentImputed.h"
#include "../include/VCFSelfProgenyImputed.h"
#include "../include/SampleManager.h"
#include "../include/KnownFamily.h"
#include "../include/Map.h"
#include "../include/option.h"
#include "../include/common.h"
#include "../include/RefCommon.h"

using namespace std;


//////////////////// LargeSelfFamilyRef ////////////////////

VCFSelfParentImputed *LargeSelfFamilyRef::create_parent_phased_vcf(
													const KnownFamily *family,
													const VCFSmall *orig_vcf,
													const VCFGeno *phased_vcf,
													const Map& gmap,
													const Option& op) {
	const auto&	fsamples = family->get_samples();
	const auto	ids = phased_vcf->extract_columns(fsamples);
	vector<string>	samples(1, family->get_mat());
	for(size_t i = 1; i < fsamples.size(); ++i) {
		if(ids[i] != string::npos)
			samples.push_back(fsamples[i]);
	}
	if(samples.size() == 1)
		return NULL;
	
	vector<string>	prog_samples(samples.begin() + 1, samples.end());
	auto	vcf_progs = VCFGeno::extract_samples(prog_samples, orig_vcf);
	auto	records = RefCommon::merge_records(phased_vcf, vcf_progs, samples);
	return new VCFSelfParentImputed(samples, records, gmap, 0.01, orig_vcf);
}

VCFSelfProgenyImputed *LargeSelfFamilyRef::create_progeny_phased_vcf(
													const KnownFamily *family,
													const VCFSmall *orig_vcf,
													const VCFGeno *ref_vcf,
													const VCFGeno *phased_vcf,
													const Map& gmap,
													const Option& op) {
	const STRVEC&	samples = family->get_samples();
	const size_t	N = samples.size();
	const auto	ids = phased_vcf->extract_columns(samples);
	vector<size_t>	prog_ids;
	for(size_t i = 1; i < N; ++i) {
		if(ids[i] != string::npos)
			prog_ids.push_back(i);
	}
	
	const size_t	phased_prog_id = prog_ids[0];
	const string&	phased_sample = samples[phased_prog_id];
	// # Keep only the first progeny
	const set<size_t>	set_prog_ids(prog_ids.begin() + 1, prog_ids.end());
	STRVEC	new_samples;
	for(size_t i = 0; i < N; ++i) {
		if(set_prog_ids.find(i) == set_prog_ids.end())
			new_samples.push_back(samples[i]);
	}
	
	STRVEC	non_phased_samples;
	for(auto p = samples.begin(); p != samples.end(); ++p) {
		if(*p != phased_sample)
			non_phased_samples.push_back(*p);
	}
	
	const auto	*non_phased_vcf = VCFGeno::extract_samples(non_phased_samples,
																	orig_vcf);
	const auto	records = RefCommon::merge_records(phased_vcf, non_phased_vcf,
																new_samples);
	const auto	ref_haps = ref_vcf->create_ref_haps();
	return new VCFSelfProgenyImputed(new_samples, records, ref_haps,
											prog_ids[0], gmap, 0.01, orig_vcf);
}

VCFGeno *LargeSelfFamilyRef::extract_parents(
									const vector<VCFSelfFillable *>& vcfs) {
	STRVEC	samples;
	for(auto p = vcfs.begin(); p != vcfs.end(); ++p) {
		samples.push_back((*p)->get_samples()[0]);
	}
	
	const size_t	M = vcfs[0]->size();
	vector<GenoRecord *>	records;
	for(size_t i = 0; i < M; ++i) {
		const ll	pos = vcfs[0]->get_record(i)->get_pos();
		vector<int>	geno;
		for(size_t j = 0; j < vcfs.size(); ++j) {
			geno.push_back(vcfs[j]->get_record(i)->get_geno(0));
		}
		auto	record = new GenoRecord(pos, geno);
		records.push_back(record);
	}
	return new VCFGeno(samples, records, vcfs[0]->get_ref_vcf());
}

bool LargeSelfFamilyRef::is_intersect(const vector<string>& v,
											const set<string>& s) {
	for(auto p = v.begin(); p != v.end(); ++p) {
		if(s.find(*p) != s.end())
			return true;
	}
	return false;
}

VCFGeno *LargeSelfFamilyRef::impute(
								const vector<const KnownFamily *>& families,
								const VCFSmall *orig_vcf,
								VCFGeno *merged_vcf,
								const VCFGeno *ref_vcf,
								const Map& gmap, const Option& op) {
	if(families.empty())
		return NULL;
	
	const auto	*phased_vcf = VCFGeno::join(merged_vcf, ref_vcf,
												orig_vcf->get_samples());
	
	vector<VCFSelfImputable *>	vcfs;
	const STRVEC&	phased_samples = phased_vcf->get_samples();
	set<string>	set_phased(phased_samples.begin(), phased_samples.end());
	for(auto p = families.begin(); p != families.end(); ++p) {
		const KnownFamily	*family = *p;
		STRVEC	samples = family->get_samples();
		if(set_phased.find(samples[0]) != set_phased.end()) {
			auto	*vcf1 = create_parent_phased_vcf(family, orig_vcf,
														phased_vcf, gmap, op);
			if(vcf1 != NULL)
				vcfs.push_back(vcf1);
		}
		else if(is_intersect(samples, set_phased)) {
			auto	*vcf2 = create_progeny_phased_vcf(family, orig_vcf, ref_vcf,
														phased_vcf, gmap, op);
			vcfs.push_back(vcf2);
		}
	}
	
	VCFSelfImputable::impute_VCFs(vcfs, op.num_threads);
	vector<const VCFGenoBase *>	vcfs1(vcfs.begin(), vcfs.end());
	if(merged_vcf != NULL)
		vcfs1.push_back(merged_vcf);
	
	if(vcfs1.empty())
		return NULL;
	
	auto	*new_vcf = VCFGeno::join(vcfs1, orig_vcf->get_samples());
	Common::delete_all(vcfs);
	return new_vcf;
}
